sandbox/cgroup/monitor.go

```diff
@@ -0,0 +1,151 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package cgroup
+
+import (
+	"errors"
+	"os"
+	"path"
+
+	"github.com/snapcore/snapd/sandbox/cgroup/inotify"
+)
+
+// appMonitorData contains all the data to monitor a specific Snap:
+// its name, the list of paths to monitor, and the channel to send
+// the notification when all the paths have been deleted.
+type appMonitorData struct {
+	name        string
+	cgroupPaths []string
+	npaths      int
+	channel     chan string
+}
+
+// CGroupMonitor allows to monitor several CGroups, detect when all
+// the running instances of each one have been closed, and notify
+// them separately. It should be considered a singleton, and
+// obtained using MonitorSingleton().
+type CGroupMonitor struct {
+	watched map[string][]*appMonitorData
+	watcher *inotify.Watcher
+	channel chan appMonitorData
+}
+
+var currentCGroupMonitor = CGroupMonitor{
+	watcher: nil,
+	channel: make(chan appMonitorData),
+	watched: make(map[string][]*appMonitorData),
+}
+
+func onFilesDeleted(filename string) {
+	basePath := path.Dir(filename)
+	appWatchers := currentCGroupMonitor.watched[basePath]
+	var newList []*appMonitorData
+	for _, app := range appWatchers {
+		for _, folder := range app.cgroupPaths {
+			if folder == filename {
+				app.npaths--
+			}
+		}
+		if app.npaths == 0 {
+			// all the folders have disappeared, so notify that this app has no more instances running
+			app.channel <- app.name
+		} else {
+			newList = append(newList, app)
+		}
+	}
+	if len(newList) != 0 {
+		currentCGroupMonitor.watched[basePath] = newList
+	} else {
+		delete(currentCGroupMonitor.watched, basePath)
+		currentCGroupMonitor.watcher.RemoveWatch(basePath)
+	}
+}
+
+func onFilesAdded(newApp *appMonitorData) {
+	if newApp.npaths == 0 {
+		newApp.channel <- newApp.name
+		return
+	}
+	addedPaths := false
+	for _, fullPath := range newApp.cgroupPaths {
+		basePath := path.Dir(fullPath) // Monitor the path containing this folder
+		_, exists := currentCGroupMonitor.watched[basePath]
+		if exists {
+			continue
+		}
+		err := currentCGroupMonitor.watcher.AddWatch(basePath, inotify.InDelete)
+		if err != nil {
+			continue
+		}
+		if _, err := os.Stat(fullPath); errors.Is(err, os.ErrNotExist) {
+			// if the file/folder to monitor doesn't exist after the parent being added, remove it
+			currentCGroupMonitor.watcher.RemoveWatch(basePath)
+			continue
+		}
+		currentCGroupMonitor.watched[basePath] = append(currentCGroupMonitor.watched[basePath], newApp)
+		addedPaths = true
+	}
+	if !addedPaths {
+		// if the files/folders to monitor don't exist, send the notification now
+		newApp.channel <- newApp.name
+	}
+}
+
+func monitorMainLoop() {
```

Comment:
this should be a method on CGroupMonitor

Reply:
Done, and also onFilesAdded() and onFilesDeleted()